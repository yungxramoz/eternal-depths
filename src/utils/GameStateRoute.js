import React from 'react'
import { useSelector } from 'react-redux'
import { Navigate, Outlet, useLocation } from 'react-router-dom'
import GAME_STATE from '../constants/game-state'
import ROUTE from '../constants/routes'

const GameStateRoute = () => {
  const gameState = useSelector((state) => state.game.gameState)
  const location = useLocation()

  const idlePaths = [ROUTE.HOME, ROUTE.NEW_CHARACTER, ROUTE.LEADERBOARD]
  const overPaths = [ROUTE.GAME_OVER]
  const wonPaths = [ROUTE.GAME_WON]
  const playingPaths = [ROUTE.DUNGEON]

  const navigateToPath = (path) => {
    return <Navigate to={path} />
  }

  if (gameState === GAME_STATE.IDLE && !idlePaths.includes(location.pathname)) {
    return navigateToPath(ROUTE.HOME)
  }
  if (gameState === GAME_STATE.OVER && !overPaths.includes(location.pathname)) {
    return navigateToPath(ROUTE.GAME_OVER)
  }
  if (gameState === GAME_STATE.WON && !wonPaths.includes(location.pathname)) {
    return navigateToPath(ROUTE.GAME_WON)
  }
  if (
    gameState === GAME_STATE.PLAYING &&
    !playingPaths.includes(location.pathname)
  ) {
    return navigateToPath(ROUTE.DUNGEON)
  }

  return <Outlet />
}

export default GameStateRoute
