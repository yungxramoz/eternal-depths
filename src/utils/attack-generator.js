import { ATTACK } from '../constants/attack-type'

const getRandomAttack = () => {
  const attacks = Object.values(ATTACK)
  return attacks[Math.floor(Math.random() * attacks.length)]
}

export const generateAttack = (attack) => {
  if (!attack) {
    attack = getRandomAttack()
  }
  return attack
}
