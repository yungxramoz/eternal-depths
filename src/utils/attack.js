import ATTACK_RESULT from '../constants/attack-result'
import {
  criticalChance,
  damagePoints,
  evasionChance,
  toHitChance,
} from './stats'

export const calculateDamage = (
  attack,
  sourceStats,
  targetStats,
  isEncounter = false,
) => {
  let damage = determineDamage(sourceStats, isEncounter)
  damage = Math.max(damage + attack.additionalDamage, 1) // additional damage can be negative
  if (!attack.safeHit) {
    const isHit = determineHit(sourceStats, targetStats)
    if (!isHit) {
      return {
        result: ATTACK_RESULT.MISSED,
        damage: 0,
      }
    }

    const isEvaded = determineEvade(targetStats)
    if (isEvaded) {
      return {
        result: ATTACK_RESULT.EVADED,
        damage: 0,
      }
    }

    const isCritical = determineCritical(sourceStats)
    if (isCritical) {
      return {
        result: ATTACK_RESULT.CRITICAL,
        damage: Math.floor(damage * 1.5),
      }
    }
  }

  return {
    result: ATTACK_RESULT.HIT,
    damage,
  }
}

const determineDamage = (sourceStats, isEncounter) => {
  const damageModifier = damagePoints(1, sourceStats.strength, isEncounter)
  const minDamage = sourceStats.minDamage + damageModifier
  const maxDamage = sourceStats.maxDamage + damageModifier

  const damage = Math.floor(Math.random() * (maxDamage - minDamage)) + minDamage

  return damage
}

const determineHit = (sourceStats, targetStats) =>
  Math.random() < toHitChance(sourceStats.precision, targetStats.agility)

const determineEvade = (targetStats) =>
  Math.random() < evasionChance(targetStats.agility)

const determineCritical = (sourceStats) =>
  Math.random() < criticalChance(sourceStats.precision)
