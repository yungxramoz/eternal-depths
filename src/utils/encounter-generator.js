import { ENCOUNTERS } from '../constants/encounter'
import { ENCOUNTER_BOSS } from '../constants/encounter-boss'

export const generateEncounter = (
  level,
  isBoss = false,
  random = true,
  encounter = null,
) => {
  if (random) {
    const encounters = isBoss ? ENCOUNTER_BOSS : ENCOUNTERS
    const randomIndex = Math.floor(Math.random() * encounters.length)
    const randomEncounter = encounters[randomIndex]
    encounter = randomEncounter
  }

  const name = encounter.name
  const fileName = encounter.fileName
  const idleAnimation = encounter.idleAnimation
  const stages = encounter.stages
  const style = encounter.style

  const points = level * 2
  const stats = { ...encounter.baseStats }
  generateStats(points, stats, level)

  const maxHp = stats.health * 4 + 10
  const hp = maxHp
  const minDamage = stats.strength * 2
  const maxDamageRange = isBoss ? 5 : 3
  const maxDamage = minDamage + maxDamageRange

  return {
    level,
    isBoss,
    name,
    fileName,
    idleAnimation,
    stages,
    style,
    stats,
    maxHp,
    hp,
    minDamage,
    maxDamage,
  }
}

export const encounterLevel = (stage) => {
  return Math.floor((stage - 1) / 5) + 1
}

const generateStats = (points, stats, level) => {
  const levelModification = level - 1
  stats.health += levelModification
  stats.strength += levelModification
  stats.agility += levelModification
  stats.precision += levelModification
  for (let i = 0; i < points; i++) {
    const randomStat = Math.floor(Math.random() * 4)
    switch (randomStat) {
      case 0:
        stats.health += 1
        break
      case 1:
        stats.strength += 1
        break
      case 2:
        stats.agility += 1
        break
      case 3:
        stats.precision += 1
        break
      default:
        break
    }
  }
}
