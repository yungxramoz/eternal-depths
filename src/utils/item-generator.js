import ARMOR_TYPE from '../constants/armor-type'
import RARITY from '../constants/rarity'
import WEAPON_TYPE from '../constants/weapon-type'

const items = [
  ARMOR_TYPE.HELMET,
  ARMOR_TYPE.ARMOR,
  ARMOR_TYPE.GLOVES,
  ARMOR_TYPE.GREAVES,
  WEAPON_TYPE.SWORD,
  WEAPON_TYPE.SHIELD,
]

export const generateItem = (name, item, rarity) => {
  item = item ? { ...item } : null
  if (!item) {
    item = { ...items[Math.floor(Math.random() * items.length)] }
  }
  if (!name) {
    name = item.type.charAt(0).toUpperCase() + item.type.slice(1)
  }
  if (!rarity) {
    const randomRarity = Math.random()
    if (randomRarity <= 0.65) {
      rarity = RARITY.COMMON
    } else if (randomRarity <= 0.85) {
      rarity = RARITY.RARE
    } else if (randomRarity <= 0.95) {
      rarity = RARITY.EPIC
    } else {
      rarity = RARITY.LEGENDARY
    }
  }

  const stats = { ...item.stats }
  generateStats(item.type, stats, rarity)
  item.stats = stats

  //TODO currently only swords are weapons
  item.type = item.type === 'sword' ? 'weapon' : item.type

  return { ...item, rarity, name }
}

const generateStats = (type, stats, rarity) => {
  let statIncrease
  switch (rarity) {
    case RARITY.COMMON:
      statIncrease = 1
      break
    case RARITY.RARE:
      statIncrease = 3
      break
    case RARITY.EPIC:
      statIncrease = 5
      break
    case RARITY.LEGENDARY:
      statIncrease = 8
      break
    default:
      break
  }

  if (type === 'weapon') {
    statIncrease = Math.floor(Math.random() * (statIncrease * 2)) + statIncrease
    stats.minDamage += Math.floor(Math.random() * statIncrease)
    stats.maxDamage += statIncrease - stats.minDamage
  } else {
    for (let i = 0; i < statIncrease; i++) {
      stats[randomStat(stats)] += 1
    }
  }
}

const randomStat = (stats) => {
  const statKeys = Object.keys(stats)
  return statKeys[Math.floor(Math.random() * statKeys.length)]
}
