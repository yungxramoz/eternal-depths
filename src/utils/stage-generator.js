export const generateStage = (encounter) => {
  const stages = encounter.stages
  const randomIndex = Math.floor(Math.random() * stages.length)
  return stages[randomIndex]
}
