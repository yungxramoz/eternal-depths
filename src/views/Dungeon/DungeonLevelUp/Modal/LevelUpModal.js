import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import Modal from '../../../../components/organisms/Modal/Modal'
import RewardButton from '../../../../components/organisms/RewardButton/RewardButton'
import RpgSeparator from '../../../../components/templates/RpgSeperator/RpgSeparator'
import { RPGUI_ICON } from '../../../../constants/rpgui-icon'
import './LevelUpModal.css'
import { generateAttack } from '../../../../utils/attack'
import { characterLearnAttack } from '../../../../store/game/gameSlice'

const LevelUpModal = ({ isOpen, setIsOpen }) => {
  const dispatch = useDispatch()
  const characterLevel = useSelector(
    (state) => state.game.character.current.level,
  )
  const characterAttacks = useSelector(
    (state) => state.game.character.current.attacks,
  )
  const [attackReward] = useState(generateAttack())

  const selectAttackReward = () => {
    dispatch(characterLearnAttack(attackReward))
    setIsOpen(false)
  }

  const selectStatReward = () => {
    console.log('Stat reward selected')
    setIsOpen(false)
  }

  return (
    <>
      {isOpen && (
        <Modal title={`Level ${characterLevel} reached!`} dismissable={false}>
          <p>Choose your reward:</p>
          <div className="reward-selection-container">
            <RewardButton
              icon={RPGUI_ICON.BOOK}
              title={attackReward.name}
              subtitle={`Cooldown: ${attackReward.cooldown}`}
              description={attackReward.description}
              disabled={characterAttacks.length >= 3}
              onClick={selectAttackReward}
            ></RewardButton>
            <RpgSeparator golden />
            <RewardButton
              icon={RPGUI_ICON.POTION}
              title="Attribute"
              description="Improve 2 attributes"
              onClick={selectStatReward}
            ></RewardButton>
          </div>
        </Modal>
      )}
    </>
  )
}

export default LevelUpModal
