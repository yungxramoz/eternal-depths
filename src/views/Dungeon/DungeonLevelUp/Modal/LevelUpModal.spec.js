import { configureStore } from '@reduxjs/toolkit'
import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import gameReducer from '../../../../store/game/gameSlice'
import { generateAttack } from '../../../../utils/attack'
import LevelUpModal from './LevelUpModal'

jest.mock('../../../../utils/attack', () => ({
  generateAttack: jest.fn(),
}))

describe('LevelUpModal', () => {
  let store
  let setIsOpen

  beforeEach(() => {
    store = configureStore({ reducer: { game: gameReducer } })
    setIsOpen = jest.fn()
    generateAttack.mockReturnValue({
      name: 'Test Attack',
      cooldown: 1,
      description: 'Test Description',
    })
  })

  it('renders correctly', () => {
    render(
      <Provider store={store}>
        <LevelUpModal isOpen={true} setIsOpen={setIsOpen} />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('LevelUpModal Open')
  })

  it('dispatches characterLearnAttack action and hides LevelUpModal when attack reward is selected', () => {
    const dispatchSpy = jest.spyOn(store, 'dispatch')
    render(
      <Provider store={store}>
        <LevelUpModal isOpen={true} setIsOpen={setIsOpen} />
      </Provider>,
    )
    fireEvent.click(screen.getByText('Test Attack'))
    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'game/characterLearnAttack' }),
    )
    expect(setIsOpen).toHaveBeenCalledWith(false)
  })
})
