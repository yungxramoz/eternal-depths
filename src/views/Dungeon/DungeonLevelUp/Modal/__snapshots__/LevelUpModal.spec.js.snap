// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LevelUpModal renders correctly: LevelUpModal Open 1`] = `
<body>
  <div>
    <div
      class="modal-container"
    >
      <div
        class="rpgui-container container framed framed-golden modal"
      >
        <div
          class="modal-header"
        >
          <h2>
            Level 1 reached!
          </h2>
          <hr
            class="separator"
          />
        </div>
        <div
          class="modal-body"
        >
          <p>
            Choose your reward:
          </p>
          <div
            class="reward-selection-container"
          >
            <button
              class="golden-button reward-button"
            >
              <div
                class="rpgui-container container framed framed-golden-2 golden-container"
              >
                <p
                  class="reward-title"
                >
                  Test Attack
                </p>
                <div
                  class="reward-button-container"
                >
                  <div
                    class="rpgui-icon book reward-icon"
                  />
                  <div
                    class="reward-button-content"
                  >
                    <p
                      class="reward-subtitle"
                    >
                      <i>
                        Cooldown: 1
                      </i>
                    </p>
                    <p
                      class="reward-description"
                    >
                      Test Description
                    </p>
                  </div>
                </div>
              </div>
            </button>
            <hr
              class="separator golden"
            />
            <button
              class="golden-button reward-button"
            >
              <div
                class="rpgui-container container framed framed-golden-2 golden-container"
              >
                <p
                  class="reward-title"
                >
                  Attribute
                </p>
                <div
                  class="reward-button-container"
                >
                  <div
                    class="rpgui-icon potion reward-icon"
                  />
                  <div
                    class="reward-button-content"
                  >
                    <p
                      class="reward-subtitle"
                    >
                      <i />
                    </p>
                    <p
                      class="reward-description"
                    >
                      Improve 2 attributes
                    </p>
                  </div>
                </div>
              </div>
            </button>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <hr
            class="separator"
          />
        </div>
      </div>
      <div
        class="modal-background"
      />
    </div>
  </div>
</body>
`;
